### ISSUE TITLE
Changes required for pip's new strict dependency resolver

### ISSUE THREAD

**Issue by hvy** [optuna/optuna#1880]
## Motivation

pip's default dependency resolved is made strict in 20.3. C.f. https://pip.pypa.io/en/stable/user_guide/#resolver-changes-2020 

Following errors should be resolved or addressed to cope with this.

```
pytorch-lightning 0.9.0 requires tensorboard==2.2.0, but you'll have tensorboard 2.3.0 which is incompatible.
tensorflow 2.3.1 requires numpy<1.19.0,>=1.16.0, but you'll have numpy 1.19.2 which is incompatible.
allennlp 1.0.0 requires spacy<2.3,>=2.1.0, but you'll have spacy 2.3.2 which is incompatible.
allennlp 1.0.0 requires torch<1.6.0,>=1.5.0, but you'll have torch 1.6.0+cpu which is incompatible.
mlflow 1.11.0 requires alembic<=1.4.1, but you'll have alembic 1.4.3 which is incompatible.
mlflow 1.11.0 requires sqlalchemy<=1.3.13, but you'll have sqlalchemy 1.3.19 which is incompatible.
```

_From https://app.circleci.com/pipelines/github/optuna/optuna/2967/workflows/fd940959-f6aa-4b74-8f34-800dbaf89d3c/jobs/59706_

## Description

See above.

## Additional context (optional)

Currently, all of the errors stem from optional dependencies (integration modules). Given that the repository of integrations modules may grow, although not planned, it may not be realistic to assume that all of them should work at the same time. This is what's somewhat assumed, at least by the CI, for the time being.

**Comment by hvy** [comment]
We may resort to keep using the old not-strict resolver until we support pip 21.0.

**Comment by crcrpar** [comment]
We've used `pip==21` despite this issue, and I've seen no failures. However, recently https://github.com/optuna/optuna/runs/1993960604?check_suite_focus=true took about 1 hour to install integration requirements. This could be attributed to that that new AllenNLP is compatible with spacy v1 and pip tries to install multiple spacy by source build.

So, as tentative measures, we could
- constrict pip version to 20
- add option of `--use-deprecated=legacy-resolver` to latest pip

**Comment by hvy** [comment]
Thanks for the heads up and the alternatives. I personally prefer being less aggressive with `--use-deprecated=legacy-resolver` if that is sufficient, since it allows us to keep testing with the latest `pip`, until we solve the underlying dependency conflicts.

**Comment by Crissman** [comment]
@hvy Looks like this was addressed with PR #1880. Ok to close?

**Comment by nzw0301** [comment]
I remove `--use-deprecated=legacy-resolver` flag and ran CI by chance. Fortunately, it took about 6min to install libraries in [the `Coverage` job](https://github.com/nzw0301/optuna/runs/2974623849?check_suite_focus=true) without error and fixed the installation error on the master branch, e.g., [this](https://github.com/optuna/optuna/runs/2973814817?check_suite_focus=true).
